### PR TITLE
fix 68: extend fuzz test to more models

### LINKED ISSUE #68: Add Tests for Malformed Data
### Problem

PBJ has not been fully tested for parsing streams of badly formed data. 

### Solution

Write unit tests and file issues as needed to fix any cases where input data could cause problems. As long as all badly formed data can not cause high resource usage like CPU or Memory and it throws a ParseException if anything goes wrong then we are all good.

Examples of badly formed data could be:

- Very deep nested messages
- Large Lists
- Large lengths for size of length encoded data
- Probably Others

Also adding fuzz testing could be a good idea to make really sure.

### PR DESCRIPTION
**Pull request by anthony-swirldslabs**
**Description**:
1. Adding about 8 more test models to run the fuzz test against. The total number of test objects from their `ARGUMENTS` is a little over 300.
2. Splitting the `integTest` and `fuzzTest` Gradle targets to run the fuzz tests with assertions off, and making the regular `test` target depend on both.
3. Minor adjustments to heuristics used in the fuzz test.

Please note that the `THRESHOLD` mechanism will be overhauled in a future PR where I'll replace the individual tests failures with gathering of statistics across all runs, and then deciding if the test as a whole should fail or not. But this is OUTSIDE OF SCOPE of this particular PR.

**Related issue(s)**:

Fixes #68 

**Notes for reviewer**:
1. The output of the test is similar as before but it now tests against more than 300 test objects.
2. The test takes a couple of minutes to run currently. I'll be looking at optimizing it a bit in a future PR.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: pbj-integration-tests/build.gradle.kts [hunk 1]
```diff
@@ -57,8 +57,9 @@ testing {
     @Suppress("UnstableApiUsage")
     suites.getByName<JvmTestSuite>("test") {
         useJUnitJupiter()
-        targets.configureEach {
+        targets.all {
             testTask {
+                useJUnitPlatform()
```
Comment (jjohannes) [inline pbj-integration-tests/build.gradle.kts:62]:
```suggestion
```

This is already set by `useJUnitJupiter()` above. See: https://docs.gradle.org/current/userguide/jvm_test_suite_plugin.html#differences_between_similar_methods_on_jvmtestsuite_and_test_task_types

File: pbj-integration-tests/build.gradle.kts [hunk 2]
```diff
@@ -71,9 +72,34 @@ testing {
...
+
+            shouldRunAfter(tasks.test)
+
+            useJUnitPlatform { includeTags("FUZZ_TEST") }
+            enableAssertions = false
+        }
     }
 }
 
+tasks.test.configure() {
```
Comment (jjohannes) [inline pbj-integration-tests/build.gradle.kts:98]:
```suggestion
tasks.test {
```

Does the same, but is more concise.

File: pbj-integration-tests/build.gradle.kts [hunk 3]
```diff
@@ -71,9 +72,34 @@ testing {
...
+    actions.clear()
```
Comment (jjohannes) [inline pbj-integration-tests/build.gradle.kts:99]: This is a bit of a nasty hack. 🤔 Have you considered using this task – `test` – for the "normal" integration tests and not add an additional `integTest` task? Because the tests are in `test` and usually it is expected that the task `test` executes the tests in that directory. I think this also leads to unexpected behavior when executing a single test from IDEA (if you care about that).

You could also consider to:
- Move all tests into a folder called `integTest` and then register a new test suite for that folder.
- (Maybe) Move the FuzzTests into a separate folder called `fuzzTest` and register a test suite for that. Then you do not need the filtering by tag.

Registering the test suites automatically sets up the tasks with the corresponding names. The Gradle configuration would look like this:

```
testing {
    suites { 
        register<JvmTestSuite>("integTest") { 
            dependencies {
                implementation(project()) 
            }
            targets.all {
                tasks.test { dependsOn(testTask) }
            }
        }
        register<JvmTestSuite>("fuzzTest") { 
            dependencies {
                implementation(project()) 
            }
            targets.all {
                testTask { enableAssertions = false }
                tasks.test { dependsOn(testTask) }
            }
        }
    }
}
```

File: pbj-integration-tests/build.gradle.kts [hunk 4]
```diff
@@ -71,9 +72,34 @@ testing {
...
+        }
+
+        tasks.register<Test>("fuzzTest") {
+            testClassesDirs = sources.output.classesDirs
+            classpath = sources.runtimeClasspath
+
+            shouldRunAfter(tasks.test)
+
+            useJUnitPlatform { includeTags("FUZZ_TEST") }
+            enableAssertions = false
```
Comment (jjohannes) [inline pbj-integration-tests/build.gradle.kts:93]: Why is this needed?

**Review by jjohannes — CHANGES_REQUESTED** [review]

File: pbj-integration-tests/build.gradle.kts (hunk 1 above)
Comment (anthony-swirldslabs) [reply pbj-integration-tests/build.gradle.kts:62]: Will remove.

File: pbj-integration-tests/build.gradle.kts (hunk 2 above)
Comment (anthony-swirldslabs) [reply pbj-integration-tests/build.gradle.kts:98]: Will change.

File: pbj-integration-tests/build.gradle.kts (hunk 3 above)
Comment (anthony-swirldslabs) [reply pbj-integration-tests/build.gradle.kts:99]: Thanks for the suggestion! That indeed looks cleaner. I'll try to implement this and see how it goes.

File: pbj-integration-tests/build.gradle.kts (hunk 1 above)
Comment (anthony-swirldslabs) [reply pbj-integration-tests/build.gradle.kts]: Addressed in latest commit.

File: pbj-integration-tests/build.gradle.kts (hunk 2 above)
Comment (anthony-swirldslabs) [reply pbj-integration-tests/build.gradle.kts]: Addressed in latest commit.

File: pbj-integration-tests/build.gradle.kts (hunk 3 above)
Comment (anthony-swirldslabs) [reply pbj-integration-tests/build.gradle.kts:98]: Unfortunately, I didn't manage to get this working. For one, all the jvm args had to be repeated in each suite definition separately, even though the `enableAssertions` is the only difference between the two. I also had to extend the configurations for the new targets from ordinary targets similar to 

https://docs.gradle.org/current/userguide/java_testing.html#sec:configuring_java_integration_tests

otherwise Gradle wouldn't run them.

But apart from these minor things, ultimately I was unable to make the Gradle script compile the generated source at `pbj-integration-tests/build/generated/source/pbj-proto/test/` and then include the compiled classes into the classpath of the `fuzzTest` target. I'm sure I missed something trivial, but due to my poor expertise in Gradle unfortunately I was unable to find a solution for this.

While I agree that the solution in the current fix where we redefine the `test` target doesn't look conventional, it's pretty straightforward from the Gradle script perspective and gets the job done as far as running both the regular integ and the new fuzz tests is concerned. However, if you know how to solve the issues I described above, I'd be happy to learn.

Comment (jjohannes) [reply pbj-integration-tests/build.gradle.kts:98]: Thanks for giving it a try. If the `generated/source/pbj-proto/test/` makes this complicated (I didn't think much about that) then I agree that we should keep things in `src/test`.

My remark on keeping the setup this way is that I would then keep using `test` to run the normal integration tests. Which means **not** do `actions.clear()` and not do an additional `tasks.register<Test>("integTest")`

If you keep things as they are now, you could remove the whole

```
    targets.all {
            testTask {
                // We are running a lot of tests 10s of thousands, so they need to run in parallel
                systemProperties["junit.jupiter.execution.parallel.enabled"] = true
                systemProperties["junit.jupiter.execution.parallel.mode.default"] = "concurrent"
                // us parallel GC to keep up with high temporary garbage creation, and allow GC to use 40% of CPU if needed
                jvmArgs("-XX:+UseParallelGC", "-XX:GCTimeRatio=90")
//                jvmArgs("-XX:+UseZGC","-XX:ZAllocationSpikeTolerance=2")
//                jvmArgs("-XX:+UseG1GC", "-XX:GCTimeRatio=90", "-XX:MaxGCPauseMillis=100")
                // Some also need more memory
                minHeapSize = "512m"
                maxHeapSize = "4096m"
            }
        }

```

block. Because `targets.all { testTask {} }` is the same as `test.tasks {}`. The `all` might suggest that you configure more than one task, but it is not the case. It only configures the task associated with the test suite, and there is always only one – https://docs.gradle.org/current/userguide/jvm_test_suite_plugin.html#test_suite_target. The ability to add a new task to a test suite is still a missing feature. I know this part is quite confusing. ☹️ 

What you probably want to do is (if the `jvmArgs ` and `memory` settings are needed? Seems to work without as well 😄):

```
tasks.withType<Test>().configureEach {
   // ...
}
```

Comment (anthony-swirldslabs) [reply pbj-integration-tests/build.gradle.kts]: Thanks for the tip! In the latest commit, I'm only adding the `fuzzTest` target and making the `test` depend on it, and it all seems to work fine, and it looks a lot nicer in the Gradle script itself. Please have a look.

**Review by jjohannes — APPROVED** [review]

File: pbj-integration-tests/build.gradle.kts (hunk 3 above)
Comment (jjohannes) [reply pbj-integration-tests/build.gradle.kts]: Perfect. Gradle part looks good to me now!

File: pbj-integration-tests/build.gradle.kts [hunk 5]
```diff
@@ -57,25 +57,33 @@ testing {
...
+    dependsOn("fuzzTest")
+}
+
+tasks.withType<Test>().configureEach {
+    // We are running a lot of tests 10s of thousands, so they need to run in parallel
+    systemProperties["junit.jupiter.execution.parallel.enabled"] = true
+    systemProperties["junit.jupiter.execution.parallel.mode.default"] = "concurrent"
+    // us parallel GC to keep up with high temporary garbage creation, and allow GC to use 40% of CPU if needed
+    jvmArgs("-XX:+UseParallelGC", "-XX:GCTimeRatio=90")
+//                jvmArgs("-XX:+UseZGC","-XX:ZAllocationSpikeTolerance=2")
```
Comment (imalygin) [inline pbj-integration-tests/build.gradle.kts:82]: please remove these commented lines if they are not intended to be used later or provide some context in which uncommenting these lines may be useful.

File: pbj-integration-tests/src/main/java/com/hedera/pbj/integration/fuzz/FuzzTest.java [hunk 6]
```diff
@@ -37,6 +37,18 @@ public static <T> FuzzTestResult<T> fuzzTest(final T object, final double thresh
         final Random random = new Random();
```
Comment (imalygin) [inline pbj-integration-tests/src/main/java/com/hedera/pbj/integration/fuzz/FuzzTest.java:37]: Please, print out the seed of this `Random` object.

Comment (anthony-swirldslabs) [reply pbj-integration-tests/src/main/java/com/hedera/pbj/integration/fuzz/FuzzTest.java:37]: Random doesn't have a getter for its seed.
I'm working on creating a Random with a fixed seed right now, but that will come in a separate PR.

File: pbj-integration-tests/build.gradle.kts (hunk 5 above)
Comment (anthony-swirldslabs) [reply pbj-integration-tests/build.gradle.kts]: Please note that I merely moved these lines from another place. However, I agree that having a commented code isn't ideal, so I removed it in the latest commit. Please have a look.

**Review by imalygin — APPROVED** [review]
